File: backend/Formulas/energia.py
import numpy as np

def calcular_energia_cinetica(masa: float, velocidad: float) -> float:
    """
    Calcula la energía cinética (K) de un objeto.

    Args:
        masa (float): La masa del objeto (kg).
        velocidad (float): La velocidad del objeto (m/s).

    Returns:
        float: La energía cinética (Julios).

    Raises:
        ValueError: Si la masa es no positiva.
    """
    if masa <= 0:
        raise ValueError("La masa debe ser positiva.")
    return 0.5 * masa * velocidad**2

def calcular_energia_potencial_elastica(constante_elastica, deformacion):
    """
    Calcula la energía potencial elástica.
    U_elastica = 0.5 * k * x^2
    """
    if constante_elastica <= 0:
        raise ValueError("La constante elástica debe ser positiva.")
    
    energia_potencial = 0.5 * constante_elastica * deformacion**2

    return {
        "energia_potencial": energia_potencial
    }

def calcular_energia_potencial_gravitatoria(masa, altura):
    """
    Calcula la energía potencial gravitatoria.
    U_g = m * g * h
    """
    g = 9.81  # Gravedad en m/s^2

    if masa <= 0:
        raise ValueError("La masa debe ser positiva.")
    if altura < 0:
        raise ValueError("La altura no puede ser negativa.")

    energia_potencial = masa * g * altura

    return {
        "energia_potencial": energia_potencial
    }

def calcular_potencia(trabajo: float, tiempo: float) -> float:
    """
    Calcula la potencia (P) dado el trabajo realizado y el tiempo.

    Args:
        trabajo (float): El trabajo realizado (Julios).
        tiempo (float): El tiempo durante el cual se realiza el trabajo (s).

    Returns:
        float: La potencia (Watts).

    Raises:
        ValueError: Si el tiempo es no positivo.
    """
    if tiempo <= 0:
        raise ValueError("El tiempo debe ser un valor positivo.")
    return trabajo / tiempo

def calcular_potencia_velocidad(fuerza: float, velocidad: float) -> float:
    """
    Calcula la potencia (P) dado una fuerza y una velocidad constante.

    Args:
        fuerza (float): La fuerza aplicada (N).
        velocidad (float): La velocidad del objeto (m/s).

    Returns:
        float: La potencia (Watts).
    """
    return fuerza * velocidad

def calcular_trabajo_energia_cinetica(masa, velocidad_inicial, velocidad_final):
    """
    Calcula el trabajo realizado y el cambio en la energía cinética.
    W = ΔK
    ΔK = K_final - K_inicial
    K = 0.5 * m * v^2
    """
    if masa <= 0:
        raise ValueError("La masa debe ser positiva.")

    # Energía cinética inicial
    energia_cinetica_inicial = 0.5 * masa * velocidad_inicial**2

    # Energía cinética final
    energia_cinetica_final = 0.5 * masa * velocidad_final**2

    # Trabajo realizado (cambio en la energía cinética)
    trabajo = energia_cinetica_final - energia_cinetica_inicial

    return {
        "trabajo": trabajo,
        "energia_cinetica_inicial": energia_cinetica_inicial,
        "energia_cinetica_final": energia_cinetica_final,
        "velocidad_inicial": velocidad_inicial,
        "velocidad_final": velocidad_final
    }

File: src/lib/physics/energia.ts
// Energía - Cálculos de física
const GRAVEDAD = 9.81;

export function simularTrabajoEnergia(masa: number, velocidad_inicial: number, velocidad_final: number) {
  if (masa <= 0) throw new Error("La masa debe ser positiva");

  const ke_inicial = 0.5 * masa * Math.pow(velocidad_inicial, 2);
  const ke_final = 0.5 * masa * Math.pow(velocidad_final, 2);
  const trabajo = ke_final - ke_inicial;

  return {
    parametros_entrada: { masa, velocidad_inicial, velocidad_final },
    resultados: {
      energia_cinetica_inicial: Number(ke_inicial.toFixed(4)),
      energia_cinetica_final: Number(ke_final.toFixed(4)),
      trabajo_neto: Number(trabajo.toFixed(4)),
    },
    message: "Cálculo de trabajo y energía cinética exitoso."
  };
}

export function simularEnergiaPotencialGravitatoria(masa: number, altura: number) {
  if (masa <= 0) throw new Error("La masa debe ser positiva");

  const ep = masa * GRAVEDAD * altura;
  const velocidad_caida = Math.sqrt(2 * GRAVEDAD * Math.abs(altura));

  return {
    parametros_entrada: { masa, altura },
    resultados: {
      energia_potencial: Number(ep.toFixed(4)),
      velocidad_equivalente_caida: Number(velocidad_caida.toFixed(4)),
    },
    message: "Cálculo de energía potencial gravitatoria exitoso."
  };
}

export function simularEnergiaPotencialElastica(constante_elastica: number, deformacion: number) {
  if (constante_elastica <= 0) throw new Error("La constante elástica debe ser positiva");

  const ep = 0.5 * constante_elastica * Math.pow(deformacion, 2);
  const fuerza = constante_elastica * Math.abs(deformacion);

  return {
    parametros_entrada: { constante_elastica, deformacion },
    resultados: {
      energia_potencial_elastica: Number(ep.toFixed(4)),
      fuerza_resorte: Number(fuerza.toFixed(4)),
    },
    message: "Cálculo de energía potencial elástica exitoso."
  };
}
